corellium/shard/src/main/kotlin/flank/corellium/shard/Internal.kt

```diff
@@ -0,0 +1,105 @@
+package flank.corellium.shard
+
+import kotlin.math.min
+
+// Stage 1
+
+/**
+ * Create flat and easy to iterate list of [Chunk] without loosing info about app and test related to test case.
+ * @receiver The [List] of [Shard.App].
+ * @return The [List] of [Chunk] which is just different representation of input data.
+ */
+internal fun List<Shard.App>.mapToInternalChunks(): List<Chunk> =
+    mutableListOf<Chunk>().also { result ->
+        forEach { app ->
+            app.tests.forEach { test ->
+                test.cases.forEach { case ->
```

Comment:
triple `forEach` nested 😄 
could not we avoid this ?

Reply:
It does look chaotic agreed.

Reply:
It's not possible to avoid nested loops in this case.

The input structure `List<Shard.App>` is designed to not store redundant information, but for sharding-calculation is necessary to change the data representation to flat.

Anyway, It's just a simple mapper so I cannot understand why you are complaining, there is no single ugly condition here, just pure function.

Reply:
We are complaining because in our opinion this code is not so readable. I think that code review is not just about reviewing conditions.